core/src/main/java/dk/alexandra/fresco/lib/lp/LPSolver.java

```diff
@@ -281,37 +297,50 @@ public LPOutput(LPTableau tableau, Matrix<DRes<SInt>> updateMatrix, List<DRes<SI
     }
   }
 
-  protected class LpState implements DRes<LpState> {
+  protected static class LpState {
 
-    public DRes<BigInteger> terminationOut;
-    private LPTableau tableau;
-    private Matrix<DRes<SInt>> updateMatrix;
-    public List<DRes<SInt>> enteringIndex;
-    public DRes<SInt> pivot;
-    public List<BigInteger> enumeratedVariables;
-    public List<DRes<SInt>> basis;
-    public DRes<SInt> prevPivot;
+    private final int iteration;
+    private final boolean terminated;
+    private final LPTableau tableau;
+    private final Matrix<DRes<SInt>> updateMatrix;
+    private final DRes<SInt> pivot;
+    private final List<BigInteger> enumeratedVariables;
+    private final List<DRes<SInt>> basis;
+    private final DRes<SInt> prevPivot;
 
-    public LpState(BigInteger terminationOut, LPTableau tableau, Matrix<DRes<SInt>> updateMatrix,
-        List<DRes<SInt>> enteringIndex, DRes<SInt> pivot, List<BigInteger> enumeratedVariables,
-        List<DRes<SInt>> basis, DRes<SInt> prevPivot) {
-      this.terminationOut = () -> terminationOut;
+    public LpState(List<DRes<SInt>> basis, Matrix<DRes<SInt>> updateMatrix,
+        DRes<SInt> pivot, int iteration, LPTableau tableau, List<BigInteger> enumeratedVariables,
+        boolean terminated, DRes<SInt> prevPivot) {
+      // Phase two protocol
+      this.iteration = iteration;
+      this.terminated = terminated;
       this.tableau = tableau;
-      this.updateMatrix = updateMatrix;
-      this.enteringIndex = enteringIndex;
-      this.pivot = pivot;
       this.enumeratedVariables = enumeratedVariables;
-      this.basis = basis;
+      this.pivot = pivot;
       this.prevPivot = prevPivot;
+      this.updateMatrix = updateMatrix;
+      this.basis = basis;
+    }
+
+    public LpState(LPTableau tableau, Matrix<DRes<SInt>> updateMatrix,
+        DRes<SInt> pivot, List<BigInteger> enumeratedVariables,
+        List<DRes<SInt>> basis, DRes<SInt> prevPivot) {
+      this(basis, updateMatrix, pivot, 0, tableau, enumeratedVariables, false, prevPivot);
+    }
+
+    public LpState(LPTableau tableau, Matrix<DRes<SInt>> updateMatrix,
+        List<DRes<SInt>> basis, DRes<SInt> pivot, int iteration) {
+      // Phase one completed
+      this(basis, updateMatrix, pivot, iteration, tableau, null, true, pivot);
     }
 
-    @Override
-    public LpState out() {
-      return this;
+    public LpState() {
```

Comment:
Any reason not to have a `public static final LpState ABORT_STATE = new LpState(null, null, null, null, -1);` or similar? It seems it would be more clear to me.